File: go/config_masking_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry_test

import (
	"strings"
	"testing"

	telemetry "github.com/provide-io/provide-telemetry/go"
)

func TestRedactedStringMasksLoggingOTLPHeaders(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Logging.OTLPHeaders = map[string]string{"Authorization": "Bearer super-secret-token"}
	s := cfg.String()
	if strings.Contains(s, "super-secret-token") {
		t.Errorf("String() leaked header value: %s", s)
	}
	if !strings.Contains(s, "****") {
		t.Errorf("String() missing mask: %s", s)
	}
}

func TestRedactedStringMasksTracingOTLPHeaders(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Tracing.OTLPHeaders = map[string]string{"X-Api-Key": "sk-1234567890abcdef"} // pragma: allowlist secret
	s := cfg.String()
	if strings.Contains(s, "1234567890abcdef") { // pragma: allowlist secret
		t.Errorf("String() leaked header value: %s", s)
	}
	if !strings.Contains(s, "****") {
		t.Errorf("String() missing mask: %s", s)
	}
}

func TestRedactedStringMasksEndpointCredentials(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Tracing.OTLPEndpoint = "https://user:p4ssw0rd@otel.example.com/v1/traces" // pragma: allowlist secret
	s := cfg.String()
	if strings.Contains(s, "p4ssw0rd") {
		t.Errorf("String() leaked password: %s", s)
	}
	if !strings.Contains(s, "****") {
		t.Errorf("String() missing mask: %s", s)
	}
}

func TestRedactedStringMasksMetricsEndpointCredentials(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Metrics.OTLPEndpoint = "https://user:s3cret99@metrics.example.com/v1/metrics" // pragma: allowlist secret
	s := cfg.String()
	if strings.Contains(s, "s3cret99") {
		t.Errorf("String() leaked metrics password: %s", s)
	}
	if !strings.Contains(s, "****") {
		t.Errorf("String() missing mask: %s", s)
	}
}

func TestRedactedStringSafeWithNoSecrets(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{ServiceName: "my-service"}
	s := cfg.String()
	if !strings.Contains(s, "my-service") {
		t.Errorf("String() missing service name: %s", s)
	}
}

func TestShortHeaderValueFullyMasked(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Logging.OTLPHeaders = map[string]string{"X-Key": "short"}
	s := cfg.String()
	if strings.Contains(s, "short") {
		t.Errorf("String() leaked short value: %s", s)
	}
	if !strings.Contains(s, "****") {
		t.Errorf("String() missing mask: %s", s)
	}
}

func TestGoStringAlsoRedacts(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Logging.OTLPHeaders = map[string]string{"Authorization": "Bearer supersecretvalue"}
	s := cfg.GoString()
	if strings.Contains(s, "supersecretvalue") {
		t.Errorf("GoString() leaked header value: %s", s)
	}
}

func TestRedactedStringMethod(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{ServiceName: "svc"}
	s := cfg.RedactedString()
	if !strings.Contains(s, "svc") {
		t.Errorf("RedactedString() missing service name: %s", s)
	}
}

func TestEndpointWithNoPassword(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Tracing.OTLPEndpoint = "https://otel.example.com/v1/traces"
	s := cfg.String()
	if !strings.Contains(s, "otel.example.com") {
		t.Errorf("String() stripped non-secret endpoint: %s", s)
	}
}

func TestMaskHeaderValueLongViaString(t *testing.T) {
	cfg := &telemetry.TelemetryConfig{}
	cfg.Logging.OTLPHeaders = map[string]string{"Authorization": "Bearer super-secret-token"}
	s := cfg.String()
	// "Bear****" should appear — first 4 chars of "Bearer super-secret-token"
	if !strings.Contains(s, "Bear****") {
		t.Errorf("expected Bear**** in masked output, got: %s", s)
	}
}

func TestEndpointWithUsernameNoPassword(t *testing.T) {
	// URL with username but no password — hasPass is false → maskEndpointURL returns raw.
	cfg := &telemetry.TelemetryConfig{}
	cfg.Tracing.OTLPEndpoint = "https://user@otel.example.com/v1/traces"
	s := cfg.String()
	if !strings.Contains(s, "otel.example.com") {
		t.Errorf("String() stripped non-secret endpoint: %s", s)
	}
}

func TestEndpointWithPasswordAndPort(t *testing.T) {
	// URL with password AND explicit port — exercises the port branch in maskEndpointURL.
	cfg := &telemetry.TelemetryConfig{}
	cfg.Tracing.OTLPEndpoint = "https://user:s3cr3t@otel.example.com:4318/v1/traces" // pragma: allowlist secret
	s := cfg.String()
	if strings.Contains(s, "s3cr3t") {
		t.Errorf("String() leaked password in port URL: %s", s)
	}
	if !strings.Contains(s, "4318") {
		t.Errorf("String() dropped port from masked URL: %s", s)
	}
}

func TestRedactConfig_MasksHeadersAndEndpoint(t *testing.T) {
	cfg := telemetry.DefaultTelemetryConfig()
	cfg.ServiceName = "my-service"
	cfg.Logging.OTLPHeaders = map[string]string{"Authorization": "Bearer supersecrettoken"} // pragma: allowlist secret
	cfg.Tracing.OTLPEndpoint = "https://user:s3cr3t@otel.example.com/traces"                // pragma: allowlist secret

	result := telemetry.RedactConfig(cfg)

	if result["service_name"] != "my-service" {
		t.Errorf("service_name not preserved: %v", result["service_name"])
	}
	logging, ok := result["logging"].(map[string]interface{})
	if !ok {
		t.Fatal("logging field is not a map")
	}
	headers, ok := logging["otlp_headers"].(map[string]string)
	if !ok {
		t.Fatal("otlp_headers is not map[string]string")
	}
	if strings.Contains(headers["Authorization"], "supersecrettoken") {
		t.Errorf("header secret leaked: %v", headers["Authorization"])
	}
	tracing, ok := result["tracing"].(map[string]interface{})
	if !ok {
		t.Fatal("tracing field is not a map")
	}
	if strings.Contains(tracing["otlp_endpoint"].(string), "s3cr3t") { // pragma: allowlist secret
		t.Errorf("endpoint password leaked: %v", tracing["otlp_endpoint"])
	}
}

func TestRedactConfig_NilEndpointPassesThrough(t *testing.T) {
	cfg := telemetry.DefaultTelemetryConfig()
	result := telemetry.RedactConfig(cfg)
	tracing, ok := result["tracing"].(map[string]interface{})
	if !ok {
		t.Fatal("tracing field is not a map")
	}
	if tracing["otlp_endpoint"] != "" {
		t.Errorf("empty endpoint should be empty string, got: %v", tracing["otlp_endpoint"])
	}
}

File: typescript/tests/config-masking.test.ts
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest';
import { redactConfig } from '../src/config';
import type { TelemetryConfig } from '../src/config';

describe('redactConfig', () => {
  it('masks OTLP header values (long value)', () => {
    const cfg = {
      otlpHeaders: { Authorization: 'Bearer super-secret-token' }, // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('super-secret-token');
    const headers = result.otlpHeaders as Record<string, string>;
    expect(headers.Authorization).toBe('Bear****');
  });

  it('masks OTLP header values (short value — fully masked)', () => {
    const cfg = {
      otlpHeaders: { 'X-Key': 'short' },
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('short');
    const headers = result.otlpHeaders as Record<string, string>;
    expect(headers['X-Key']).toBe('****');
  });

  it('masks endpoint credentials in URL userinfo', () => {
    const cfg = {
      otlpEndpoint: 'https://user:p4ssw0rd@otel.example.com', // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('p4ssw0rd');
    expect(result.otlpEndpoint as string).toContain('****');
    expect(result.otlpEndpoint as string).toContain('user');
    expect(result.otlpEndpoint as string).toContain('otel.example.com');
  });

  it('leaves endpoint unchanged when no password present', () => {
    const cfg = {
      otlpEndpoint: 'https://otel.example.com/v1/traces',
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.otlpEndpoint).toBe('https://otel.example.com/v1/traces');
  });

  it('leaves endpoint unchanged when it is not a valid URL', () => {
    const cfg = {
      otlpEndpoint: 'not-a-url',
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.otlpEndpoint).toBe('not-a-url');
  });

  it('is safe with no secrets', () => {
    const cfg = { serviceName: 'my-app' } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.serviceName).toBe('my-app');
  });

  it('does not modify otlpHeaders when the field is absent', () => {
    const cfg = { serviceName: 'svc' } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect('otlpHeaders' in result).toBe(false);
  });

  it('does not modify otlpEndpoint when the field is absent', () => {
    const cfg = { serviceName: 'svc' } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    // otlpEndpoint is optional; when not in cfg it should not appear in result
    expect(result.otlpEndpoint).toBeUndefined();
  });

  it('does not modify empty otlpHeaders object', () => {
    const cfg = {
      otlpHeaders: {},
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    // empty headers object passes through unchanged (no keys to mask)
    expect(result.otlpHeaders).toEqual({});
  });

  it('masks exactly 8-char header value — shows first 4', () => {
    const cfg = {
      otlpHeaders: { 'X-Key': '12345678' },
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    const headers = result.otlpHeaders as Record<string, string>;
    expect(headers['X-Key']).toBe('1234****');
  });

  it('masks per-signal logs headers', () => {
    const cfg = {
      otlpLogsHeaders: { Authorization: 'Bearer secret-logs-token' }, // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('secret-logs-token');
    const headers = result.otlpLogsHeaders as Record<string, string>;
    expect(headers.Authorization).toBe('Bear****');
  });

  it('masks per-signal traces headers', () => {
    const cfg = {
      otlpTracesHeaders: { 'X-Api-Key': 'traces-secret-key' }, // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('traces-secret-key');
    const headers = result.otlpTracesHeaders as Record<string, string>;
    expect(headers['X-Api-Key']).toBe('trac****');
  });

  it('masks per-signal metrics headers', () => {
    const cfg = {
      otlpMetricsHeaders: { Authorization: 'Bearer metrics-token' }, // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('metrics-token');
    const headers = result.otlpMetricsHeaders as Record<string, string>;
    expect(headers.Authorization).toBe('Bear****');
  });

  it('does not mask per-signal headers when field is absent', () => {
    const cfg = { serviceName: 'svc' } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.otlpLogsHeaders).toBeUndefined();
    expect(result.otlpTracesHeaders).toBeUndefined();
    expect(result.otlpMetricsHeaders).toBeUndefined();
  });

  it('does not mask per-signal headers when field is empty object', () => {
    const cfg = {
      otlpLogsHeaders: {},
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.otlpLogsHeaders).toEqual({});
  });

  it('masks per-signal logs endpoint credentials', () => {
    const cfg = {
      otlpLogsEndpoint: 'https://user:p4ssw0rd@logs.example.com', // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('p4ssw0rd');
    expect(result.otlpLogsEndpoint as string).toContain('****');
  });

  it('masks per-signal traces endpoint credentials', () => {
    const cfg = {
      otlpTracesEndpoint: 'https://user:p4ssw0rd@traces.example.com', // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('p4ssw0rd');
    expect(result.otlpTracesEndpoint as string).toContain('****');
  });

  it('masks per-signal metrics endpoint credentials', () => {
    const cfg = {
      otlpMetricsEndpoint: 'https://user:p4ssw0rd@metrics.example.com', // pragma: allowlist secret
    } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(JSON.stringify(result)).not.toContain('p4ssw0rd');
    expect(result.otlpMetricsEndpoint as string).toContain('****');
  });

  it('does not mask per-signal endpoints when fields are absent', () => {
    const cfg = { serviceName: 'svc' } as Partial<TelemetryConfig> as TelemetryConfig;
    const result = redactConfig(cfg);
    expect(result.otlpLogsEndpoint).toBeUndefined();
    expect(result.otlpTracesEndpoint).toBeUndefined();
    expect(result.otlpMetricsEndpoint).toBeUndefined();
  });
});
